File: c++/c20_2_disjoint_paths.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;
#include "c20_1_ford_fulkerson_algorithm.h"

namespace {
	vector<vector<ll>> edge_disjoint(const vector<vector<ll>>& adj, ll s, ll e) {
		ll n = ssize(adj);
		vector<vector<pair<ll, ll>>> adj2(n);
		for (ll i = 1; i < n; ++i) {
			for (auto j : adj[i]) {
				adj2[i].emplace_back(j, 1);
			}
		}
		auto adj3 = edmonds_karp(adj2, s, e);
		vector<vector<ll>> ret(adj2tc(adj3, e));
		for (auto& r : ret) {
			r.push_back(s);
			while (r.back() != e) {
				for (ll i : adj[r.back()]) {
					if (adj3[i][r.back()] > 0) {
						adj3[i][r.back()] = 0;
						r.push_back(i);
						break;
					}
				}
			}
		}
		return ret;
	}

	vector<vector<ll>> node_disjoint(const vector<vector<ll>>& adj, ll s, ll e) {
		ll l = ssize(adj);
		vector<vector<ll>> adj2(adj.size());
		for (ll i = 0; i < l; ++i) {
			if (i != s && i != e) {
				adj2[i].push_back(i + l);
			}
		}
		adj2.insert(adj2.end(), adj.begin(), adj.end());
		vector<vector<ll>> ret;
		for (const auto& r : edge_disjoint(adj2, s + l, e)) {
			ret.resize(ret.size() + 1);
			ret.back().push_back(s);
			for (ll i = 1; i < ssize(r); i += 2)
				ret.back().push_back(r[i]);
		}
		return ret;
	}

	void test_edge(const vector<vector<ll>>& adj, ll s, ll e, ll l) {
		set<pair<ll, ll>> edges;
		for (ll i = 0; i < ssize(adj); ++i) {
			for (ll j : adj[i]) {
				edges.emplace(i, j);
			}
		}
		auto ret = edge_disjoint(adj, s, e);
		EXPECT_EQ(ret.size(), l);
		for (const auto& r : ret) {
			for (auto p : r | views::pairwise) {
				EXPECT_TRUE(edges.contains(p));
				edges.erase(p);
			}
		}
	}

	void test_node(const vector<vector<ll>>& adj, ll s, ll e, ll l) {
		auto nodes = views::iota(1ll, ll(ssize(adj))) | ranges::to<unordered_set>();
		nodes.erase(s);
		nodes.erase(e);
		auto ret = node_disjoint(adj, s, e);
		EXPECT_EQ(ret.size(), l);
		for (const auto& r : ret) {
			for (ll i = 1; i < ssize(r) - 1; ++i) {
				EXPECT_TRUE(nodes.contains(r[i]));
				nodes.erase(r[i]);
			}
		}
		for (const auto& r : ret) {
			for (auto [a, b] : r | views::pairwise) {
				EXPECT_TRUE(ranges::find(adj[a], b) != adj[a].end());
			}
		}
	}
}

TEST(C202DisjointPaths, edgeDisjoint) {
	test_edge({{},{2},{}}, 1, 2, 1);
	test_edge({{},{2,4},{4},{2,5,6},{3,5},{6},{}}, 1, 6, 2);
}

TEST(C202DisjointPaths, nodeDisjoint) {
	test_node({{},{2},{}}, 1, 2, 1);
	test_node({{},{2,4},{4},{2,5,6},{3,5},{6},{}}, 1, 6, 1);
}

File: python/c20_2_disjoint_paths.py
import unittest
from itertools import pairwise
from c20_1_ford_fulkerson_algorithm import ford_fulkerson,adj2tc

def edge_disjoint(adj,s,e):
    adj2=ford_fulkerson([[(j,1) for j in i] for i in adj],s,e)
    for _ in range(adj2tc(adj2,e)):
        r=[s]
        while r[-1]!=e:
            i=next(i for i in adj[r[-1]] if adj2[i][r[-1]]>0)
            adj2[i][r[-1]]=0
            r.append(i)
        yield r

def node_disjoint(adj,s,e):
    l=len(adj)
    for r in edge_disjoint([[] if i in (s,e) else [i+l] for i in range(l)]+adj,s+l,e):
        yield [s,*r[1::2]]

class Test(unittest.TestCase):
    def assertEdgeDisjoint(self,adj,s,e,l):
        with self.subTest(adj=adj,s=s,e=e):
            edges = {(i,j) for i,a in enumerate(adj) for j in a}
            ret = [*edge_disjoint(adj,s,e)]
            self.assertEqual(len(ret),l)
            for r in ret:
                for p in pairwise(r):
                    self.assertIn(p,edges)
                    edges.discard(p)
    def assertNodeDisjoint(self,adj,s,e,l):
        with self.subTest(adj=adj,s=s,e=e):
            nodes = {*range(1,len(adj))}-{s,e}
            ret=[*node_disjoint(adj,s,e)]
            self.assertEqual(len(ret),l)
            for r in ret:
                for a in r[1:-1]:
                    self.assertIn(a,nodes)
                    nodes.discard(a)
            for r in ret:
                for a,b in pairwise(r):
                    self.assertIn(b,adj[a])
    def test(self):
        adj=[[],[2],[]]
        self.assertEdgeDisjoint(adj,1,2,1)
        self.assertNodeDisjoint(adj,1,2,1)

        adj=[[],[2,4],[4],[2,5,6],[3,5],[6],[]]
        self.assertEdgeDisjoint(adj,1,6,2)
        self.assertNodeDisjoint(adj,1,6,1)

if __name__=='__main__':
    unittest.main()
